corfudb/src/main/java/site/ycsb/db/CorfuDBClient.java

```diff
@@ -0,0 +1,261 @@
+    executorService.submit(() -> {
+        try {
+          while (true) {
+            CorfuRuntime cpRt = new CorfuRuntime(connectionStr).connect();
+            CorfuTable<String, byte[]> tableToCheckpoint = cpRt.getObjectsView()
+                .build()
+                .setTypeToken(new TypeToken<CorfuTable<String, byte[]>>() {
+                })
+                .setStreamName(TABLE_NAME)
+                .open();
+
+            long start = System.currentTimeMillis();
+            MultiCheckpointWriter<CorfuTable> mcw = new MultiCheckpointWriter<>();
+            mcw.addMap(tableToCheckpoint);
+            Token trimPoint = mcw.appendCheckpoints(cpRt, "Checkpointer-thread");
+            long end = System.currentTimeMillis();
+            long sleepTime = 1000 * 60 * 2;
+            System.out.println(">>>> Checkpointer took " + (end - start) + " ms sleeping for " + sleepTime + "seconds");
+            Thread.sleep(sleepTime);
+            cpRt.getAddressSpaceView().prefixTrim(trimPoint);
+            cpRt.getAddressSpaceView().gc();
+            cpRt.shutdown();
+          }
+        } catch (Exception e) {
+          System.out.println("Checkpointer encountered an error! " + e);
+          Runtime.getRuntime().halt(-1);
+        }
+      });
+  }
+
+  @Override
+  public void init() throws DBException {
+    Properties props = getProperties();
+
+    boolean singleMode = Boolean.parseBoolean(props.getProperty(CLIENT_MODE, "false"));
+
+    if (THREAD_COUNT.incrementAndGet() == 1) {
+      initCheckpointer(props.getProperty(CONNECTION_STRING));
+    }
+
+    synchronized (THREAD_COUNT) {
+      if (singleMode) {
+        // Share the same CorfuRuntime across workers
+        if (singletonRt == null) {
+          singletonRt = new CorfuRuntime(props.getProperty(CONNECTION_STRING)).connect();
+        }
+
+        corfuRuntime = singletonRt;
+      } else {
+        // Create a new CorfuRuntime per worker
+        corfuRuntime = new CorfuRuntime(props.getProperty(CONNECTION_STRING)).connect();
+      }
+    }
+
+    // Since NO_CACHE option is not specified this will retrieve one table instance per CorfuRuntime
+    userTable = corfuRuntime.getObjectsView()
+        .build()
+        .setTypeToken(new TypeToken<CorfuTable<String, byte[]>>() {
+        })
+        .setStreamName(TABLE_NAME)
+        .open();
+
+    long start = System.currentTimeMillis();
+    int size = userTable.size();
+    System.out.println("Table: " + TABLE_NAME + " size: " + size + " ref: " + System.identityHashCode(userTable)
+        + " loadTime: " + (System.currentTimeMillis() - start));
+    LOGGER.info("connected!");
+  }
+
+  @Override
+  public void cleanup() throws DBException {
+    System.out.println("[Table] " + TABLE_NAME + " size " + userTable.size()
+        + " " + System.identityHashCode(userTable));
+    if (THREAD_COUNT.decrementAndGet() == 0) {
+      corfuRuntime.shutdown();
```

Comment:
This needs to account for the case where CorfuRuntime is not shared, which in this case will not shutdown other runtimes.